src/project/internal/thumbnailcreator.cpp

```diff
@@ -0,0 +1,29 @@
+#include "thumbnailcreator.h"
+
+using namespace au::project;
+
+void ThumbnailCreator::onThumbnailCreated(bool success)
+{
+    m_thumbnailCreated.send(success);
+}
+
+muse::async::Channel<std::string> ThumbnailCreator::captureThumbnailRequested() const
+{
+    return m_createThumbnailRequested;
+}
+
+muse::Ret ThumbnailCreator::createThumbnail(const std::string& path)
+{
+    muse::Ret ret;
+    QEventLoop loop;
+    m_thumbnailCreated.resetOnReceive(this);
```

Comment:
I think it's better to move it after `loop.exec()`.
There may be a potential problem here, in `onReceive` we pass a lambda in which we save a reference to a local  variable (loop). If someone for some reason calls `onThumbnailCreated`, and `onReceive` is triggered, then there may be a crash. Therefore, it is safer for us to simply delete this callback after it is triggered.

Reply:
Good catch. Done.